docs/src/userguide/freethreading.rst

```diff
@@ -160,3 +160,297 @@ to require extreme care to make it work correctly).
+          for word in words:
+            overall_result.add(word)
+
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+      for _ in executor.map(read_from_file, filenames):
+        pass
+    return overall_result
+
+The less your threads interact, the less chance there is for bugs, the less
+need there is for locking to control their interaction, and the less likely
+they are to slow each other down by invaliding the CPU cache for other
+threads.
+
+Should you use ``prange``?
+--------------------------
+
+Although ``prange`` is the parallelization mechanism *built in* to Cython, it
+is not the only option, and probably should not be your default option.
+
+``prange`` is a fairly thin wrapper over OpenMP's "parallel for".  This means
+it is ideal for problems where you have a big loop, every iteration is basically
+the same, and the result of each iteration is independent of any other iteration.
+If this does not describe your problem then ``prange`` is probably not the
+solution.
+
+Remember that all the threading options available in Python are also available
+in Cython.  For example you can start threads with ``threading.Thread`` or
+``concurrent.futures.ThreadPoolExecutor`` and this is much more flexible than
+``prange``.  Similarly the synchronization tools in ``threading.Thread``
+are also available in Cython.
+
+Try to avoid Python code in ``prange``
+--------------------------------------
+
+``prange`` has some slightly unintuitive behaviour about which data is
+shared and which isn't.  Typically C variables (e.g. ``int``, ``float``) are
+treated as "thread-local" and so each thread has its own copy. However,
+Python object variables are treated as shared between all the threads.
+
+This means that::
+
+  cdef int i
+  cdef int total = 0
+  for i in cython.parallel.prange(10, nogil=True):
+    tmp = i**2
+    total += tmp
+
+should work fine - each thread has its own ``tmp`` and ``total`` is
+a "reduction" (so treated in a thread-safe manner).  However::
+
+  cdef int i
+  cdef int total = 0
+  cdef object tmp
+  for i in cython.parallel.prange(10, nogil=True):
+    with gil:
+      tmp = i**2
+      total += tmp
+
+In this case there is only a single value of ``tmp`` shared between all the threads, and
+all are continuously overwriting each other's values.  Additionally Cython does not
+currently ensure that ``tmp`` is even reference-counted correctly so you are at risk
+of crashes or memory-leaks in addition to getting a nonsense answer.
+
+If you do want to work with Python objects then it is best to move them into
+a function and just have the loop call the function::
+
+  cdef int square(int x):
+    cdef object tmp = x**2
+    return tmp
+
+  # ...
+
+  cdef int i
+  cdef int total = 0
+  for i in cython.parallel.prange(10, nogil=True):
+    with gil:
+      total += square(i)
+
+Since ``tmp`` is now local to the function scope, each function call has its own copy
+and thus there is no conflict between threads.
```

Comment:
I missed the bit that the function return value is safe because it's converted to a C `int`. Let's make that clearer.
```suggestion
  cdef int square(int x):
    cdef object tmp = x**2
    cdef int result = tmp
    return result

  # ...

  cdef int i
  cdef int total = 0
  for i in cython.parallel.prange(10, nogil=True):
    with gil:
      total += square(i)

Since ``tmp`` is now local to the function scope, each function call has its own copy
and thus there is no conflict of Python objects between threads.
```